### DISCUSSION TITLE
Better performance with multithreading binary content comparison

### DISCUSSION THREAD

**Discussion by xerces8** [WinMerge/winmerge#1920] [Ideas]
When comparing two folders I noticed that the default "Full Contents" method is about twice as fast the "Binary Contents" method, even if the docs say: This method is faster than the Quick contents compare method (which in turn is: This method is faster than Full Contents, because it does not load the files.)

The reason is probably that the former use multiple CPU cores/threads, while the binary mode uses one.

For binary contents method Task manager shows almost no load on my SSD, while about 33% load on the CPU (i7-9850H 6core/12thread) - mostly by MsMpEng.exe

Some quick benchmarks (comparing two big folders, that happen to be identical):

```
default options (Full Contents)      8:12     100% CPU, 4% SSD
Binary mode                         16:20     almost zero % CPU , MsMpEng.exe uses 6% CPU , 4% SSD        total CPU: about 33%
Quick Contents                       6:06     50-100 % CPU , 0-10 % SSD
Kdiff3                              15:04     single thread, low SSD usage
```

Is there a reason for not using multiple threads for binary contents method? It may not give more speed on HDD, but today we use SSD. Also with a lot of RAM, there is a good chance the files are in cache, even when using a HDD.

PS: I used WinMergePortable-2.16.30  64 bit version

**Comment by sdottaka** [comment]
In the Binary Contents compare method, single threading was faster in my environment, so I made it single threaded.
However, there are exceptions, enabling additional properties such as MD5, SHA-256, etc. slows down single-threading.
Do you have additional properties enabled?

**Reply by xerces8** [reply]
I just used the defaults.
Where can I find those additional properties? In Options / Compare : General, Folder and Binary I don't see anything looking like MD5 or SHA-256.

**Reply by sdottaka** [reply]
You can display the Additional Properties window by pressing the Additional Properties button in the Display Columns window displayed by the [Tools]->[Customize Columns...] menu item.

![image](https://github.com/WinMerge/winmerge/assets/98126/c9e9cd05-7833-470b-b5a1-5a4f0b405172)

**Comment by xerces8** [comment]
>  Do you have additional properties enabled?

No, none.

**Comment by xerces8** [comment]
I repeated the test with a subset of the files from first test (now I have about 2x10 GB of files in two folders that I compare, it was 2x100GB in first test) and now I get different results. Binary Content is fastest:
```
Binary Content  25 seconds   40 % CPU  100 % SSD (almost 2 GB/s)
Quick Contents  45 seconds  100 % CPU    1 GB/s SSD
Full Contents   56 seconds  100 % CPU  600 MB/s SSD
```

Strange...
